### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by AlephNotation** [bmoore20/habs#1]
I've looked it over. Cool project!

Where do would you like to start?

**Comment by bmoore20** [comment]
**4.20.21 Meeting Important Points**

**To do:**
- Include `device` in _train.py_
- Make the `model`, `optimizer`, and # of `epochs` configurable (arguments)
- Abstract and pull out code for train and test loops
- Load and mount images into Google Drive directory
- Install necessary requirements 
- Run the code! (and fix any bugs/errors that pop up)

**Notes:**
- Use Colab as a place to experiment 
- When changes are made to my local copy of the repo (PyCharm), push it up to remote GitHub repo ... then re-pull the code into the mounted Google Drive directory (reclone?).

**Links:**
- https://colab.research.google.com/github/googlecolab/colabtools/blob/master/notebooks/colab-github-demo.ipynb#scrollTo=Rmai0dD30XzL
- https://github.com/drillan/jupyter-black

**Comment by bmoore20** [comment]
**4.27.21 Meeting Notes**

- Discussed how I need to improve the current construction of the training lap
  - break up training into `training_lap` and `validation_lap`
  - `training_lap` -> how you update model weights, optimizer and back propagation
  - `validation_lap` -> used to see how good the hyperparameters are, no optimizer/back prop
  - calculate the loss for both the `training_lap` and `validation_lap` 
  - if training loss is down and the validation loss goes up, then that could be a sign of overfitting 
  - validation will let us introduce `learn rate optimizers
- Set up the `Black` code formatter with the `HABs` repo

**Comment by bmoore20** [comment]
@AlephNotation I read a little more on why we need the validation set because it is still fairly new to me. With your explanation you gave me today and what I read, it is starting to make sense. Below is my current understanding of the next steps that I should take. In order to make sure that we are on the same page before I make changes in PR #39, can you please have a quick look to check to see if my logic is correct?

Split the original HABs dataset of images into a total of 3 subsets:

1. training set (60%)
2. validation set (20%)
3. test set (20%)

The training set and validation set will both be inside the epoch loop and passed into either `training_lap` or `validation_lap` (which I still have to make). The test set will be passed into my current `evaluate` method, which run after all of the training is completed and calls the final chosen model. So I am going to keep the current evaluation method I have, but also create two new methods `training_lap` and `validation_lap`. The `validation_lap` does not perform back propagation or have an optimizer, but it does calculate loss. Both methods will return a loss. 

Thanks! Also - sorry I was a little discombobulated on the call today

**Comment by bmoore20** [comment]
**5.4.21 Meeting Notes**

- In training lap `for epoch in range(epoch):` log the **epoch #** and the **training loss** to a 7-point decimal
- Save the model 
- Give batch size to `DataLoader` object and do `for batch in data_loader:` in _training_helper.py_
- Configure batch parameter
- Give device to DataLoader??
- Helper function that creates a temp directory and takes a random 20% of the data paths from the test set to make a validation set 
- Will eventually need to use `randomCrop` on images ... maybe use a progressive GAN?
- Need to fix current model architecture to handle **3 classifications** 
- **RUN THE PROGRAM!!**

**Comment by bmoore20** [comment]
**5.11.21 Meeting Notes**

- Research ways to improve the architecture of my model -> Google the following things ...
  - image classification tasks
  - res nets -> look at architecture and how it works
  - batch norms
  - pooling
  - different types of norms
  - list of layers ... convolutional layers
  - auto-encoder, k-means
- Start messing around with different combinations of various transforms
- Set up Tensor Board
   https://colab.research.google.com/github/tensorflow/tensorboard/blob/master/docs/tensorboard_in_notebooks.ipynb
- Add `__repr__` to my custom Transformation classes so they can be represented and recorded when "print" is called

**Comment by bmoore20** [comment]
**5.18.21 Meeting Notes**

- Answered question about using `__call__` or `forward` for **transforms**
- Helped created new log file each time by adding date and time into the log file name
- Discussed IBM challenge ... https://developer.ibm.com/callforcode/?utm_content=000039JL&utm_term=10008917&p1=PSocial&p2=297608652&p3=142596057&dclid=CKX2lLvixvACFRpLDQod5ToGNw

**Comment by bmoore20** [comment]
**6.1.21 Meeting Notes**

- The problem that I was running into with getting TensorBoard to work in Colab is that I was trying to get it to run on the _local host_ for my laptop. I needed to get TensorBoard to run on the _local host_ of the Colab machine. 
- In order to do this I needed to use **magic** commands (%)
- Magic commands are special commands that help with running code in your notebook
- When you run a command with **!**, it directly executes a bash command in a subshell
- When you run a command with **%**, it executes one of the magic commands defined in IPython 
- Colab -> built on html and js 
- **QUESTION -> How do I know when to use ! (bash) vs % (magic) when doing a command in Colab?
- Make a new notebook that doesn't have a GPU that is running the TensorBoard so you can have it next to the Colab notebook that is actually running the code -> more convenient but not necessary 
-  This will work because the TensorBoard works from a directory reference ... doesn't have to be in same Colab notebook as code 
- Make a tiny function, `makeWriter` that handles functionality for creating path for writer
```
from datetime import datetime
from pathlib import Path

now = datetime.now()
now_str = now.strftime("YYYY-mm-dd_HH-MM-SS")
RUN_DIR=PATH(TENSORBOARD_DIR / now_str
```
https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter

- Colab notebook is an instance. When we close out of it, we lose our history. That is why we have to reload in the HABs github directory every time. Therefore, we want to write to the Google Drive dir. 
- ResNets -> skip it
- Replace convolution layers with ResNet layers
- People are moving away from Batch Norms (article from Ty)

**Comment by bmoore20** [comment]
Looks like someone is already doing a similar program for HABs:
https://www.noaa.gov/what-is-harmful-algal-bloom
(look under tab 4 - Where HABs Happen)

**Comment by bmoore20** [comment]
**6.10.21 Meeting Notes**

- Run the program!!
- Try running the program on a different data set ... not my HABs one (ex. ImageNet)
  - Help us know that our model is actually working properly
- It happens all the time that models don't converge
- If model does not work, we will figure it out ... we can do some crazy sh**
- **Transfer learning** -> **Pre-training**
  - drastically reduce data
  - get similar data -> train data in that domain -> freeze middle
- `Few-shot` & `One-shot` => GOOGLE TERMS
  - how to do training with no data
  - ex) Car crash -> predict rare events

**Comment by bmoore20** [comment]
**No meeting on 6.15.21**

**Comment by bmoore20** [comment]
**6.22.21 Meeting Notes**

- Talked through error below...
  - a sample: [512, 750, 3] -> [height, width, rgb]
  - a batch of 1: [1, 512, 750, 3]
  - a batch of 8: [8, 512, 750, 3]
<img width="740" alt="Screen Shot 2021-06-30 at 10 33 55 PM" src="https://user-images.githubusercontent.com/67593919/124056220-51f19800-d9f3-11eb-88f7-1c3c5fb12f23.png">
 
- Bigger batch sizes are better -> multiple samples to do back-prop
- Nerds -> pick batches in powers of 2 -> 2, 4, 8, 16, 32...
- ResNet Model and Convolutions
  - each convolution adds a different feature!!!
  - create features from data
  - each block of ResNet enhances each original image
  - each down-stream block added from previous block -> CUMULATIVE!
  - Ex. 
    - 1st layer -> edges
    - 2nd layer -> shapes 
    - 3rd layer -> dog
  - Downsampling (pooling layer)
    - each level -> pull out more abstract image
    - downsample -> more abstract

**Comment by bmoore20** [comment]
**6.29.21 Meeting Notes**

- Model is not training ... 50% is best (not better than probability) 
- Probably due to bad dataset
- If we can use model out of box and avoid bullsh**, then that's best ... doesn't look like that is the case
- Want to make the problem as simple as possible!!
  - Switch to two classes -> reduce classes from 3 to 2
     1. BGA
     2. Non-Algae
  - Will have to slightly change format of HABs Dataset
  - Does our data have enough signal to do it well?
- Also -> **Pre-training** / **Transfer learning**
  - trained network does not have enough data -> maybe not at college to get more data or maybe there is no algae in lake 
  - take adjacent dataset -> collect bga and pictures of lakes, can you find the difference between the two 
  - freeze intermediate layers (except input and outputs)
  - teach model clear & unclear, algae & clear in general -> algae detection model
  - fine tune this algae detection model -> give it very specific data for specific use case
- Train model on clear lake water
  - reconstruct clear water pictures -> Healthy Lake
  - switch problem up -> healthy lake -> no bga
- Semi-supervised image classification -> Papers With Code
   https://paperswithcode.com/task/semi-supervised-image-classification
   - **Mix-Up** & **Cut-Up**
   - BGA Image and Clear Image -> create a new image
   - .9 of orig 
   - add 1-.9 of bga image 
   - cat/dog example
   - what is new classification? -> pre-training -> make sure model is predicting 50% cat, 50% dog
   - linear combination of image vector
   - pre-train using self-supervised -> take pre-trained model and use for classification task
- ****Cut-Mix**
- HOW to integrate pre-trained model into model?!
  - in forward
  - pytorch -> freeze weights 
     https://discuss.pytorch.org/t/how-the-pytorch-freeze-network-in-some-layers-only-the-rest-of-the-training/7088/2
  - `param.requires_grad = False`-> call back-prop won't update
- Every class should have **10,000 images**!!
- Scientist -> try to keep everything the same as much as possible
- overfit model -> mix-up dataset again -> another full training loop
- CHECKPOINT mode 
  - save every 5 epochs incase your program ends in the middle of training, that way you don't loose the progress of the trained weights
  - you can then pick-up where you left off by passing in this model object into training loop 
  - load in model weights
  - freeze model -> import model -> different layers
- Might be a good idea for me to get **Colab Pro** 
  - $10 a month 
  - faster and bigger GPU
  - can run models for up to 24 hours
- GPU -> vector algebra -> designed todo graphics, happens that nn math is the same as graphics math
- TPU -> made more specifically for ml, f*** graphics
- **Training loss** is bigger than **validation loss** because training dataset has WAY more images than validation training set
  - proportional to # of images in dataset
  - if training set is 2x larger, then loss should be 2x larger
  - they should swoop in and get close to each other 
- It's okay to use batch norm here instead of dropout
- READ PAPERS that discuss current ways that people are detecting images of HABs using ML!!
